extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/results/ExportResultsFrameProcessor.java

```diff
@@ -0,0 +1,172 @@
+import org.apache.druid.java.util.common.Unit;
+import org.apache.druid.java.util.common.granularity.Granularities;
+import org.apache.druid.java.util.common.guava.Sequence;
+import org.apache.druid.msq.counters.ChannelCounters;
+import org.apache.druid.msq.util.SequenceUtils;
+import org.apache.druid.segment.BaseObjectColumnValueSelector;
+import org.apache.druid.segment.ColumnSelectorFactory;
+import org.apache.druid.segment.Cursor;
+import org.apache.druid.segment.VirtualColumns;
+import org.apache.druid.segment.column.RowSignature;
+import org.apache.druid.sql.http.ResultFormat;
+import org.apache.druid.storage.StorageConnector;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ExportResultsFrameProcessor implements FrameProcessor<Object>
+{
+  private final ReadableFrameChannel inputChannel;
+  private final ResultFormat exportFormat;
+  private final FrameReader frameReader;
+  private final StorageConnector storageConnector;
+  private final ObjectMapper jsonMapper;
+  private final int partitionNumber;
+  private final int workerNumber;
+  private final ChannelCounters channelCounter;
+
+  public ExportResultsFrameProcessor(
+      ReadableFrameChannel inputChannel,
+      ResultFormat exportFormat,
+      FrameReader frameReader,
+      StorageConnector storageConnector,
+      ObjectMapper jsonMapper,
+      int partitionNumber,
+      int workerNumber,
+      ChannelCounters channelCounter
+  )
+  {
+    this.inputChannel = inputChannel;
+    this.exportFormat = exportFormat;
+    this.frameReader = frameReader;
+    this.storageConnector = storageConnector;
+    this.jsonMapper = jsonMapper;
+    this.partitionNumber = partitionNumber;
+    this.workerNumber = workerNumber;
+    this.channelCounter = channelCounter;
+  }
+
+  @Override
+  public List<ReadableFrameChannel> inputChannels()
+  {
+    return Collections.singletonList(inputChannel);
+  }
+
+  @Override
+  public List<WritableFrameChannel> outputChannels()
+  {
+    return Collections.emptyList();
+  }
+
+  @Override
+  public ReturnOrAwait<Object> runIncrementally(IntSet readableInputs) throws IOException
+  {
+    if (readableInputs.isEmpty()) {
+      return ReturnOrAwait.awaitAll(1);
+    }
+
+    if (inputChannel.isFinished()) {
+      return ReturnOrAwait.returnObject(Unit.instance());
+    } else {
+      exportFrame(inputChannel.read());
+      return ReturnOrAwait.awaitAll(1);
+    }
+  }
+
+  private void exportFrame(final Frame frame) throws IOException
+  {
+    final RowSignature signature = frameReader.signature();
+
+    final Sequence<Cursor> cursorSequence =
+        new FrameStorageAdapter(frame, frameReader, Intervals.ETERNITY)
+            .makeCursors(null, Intervals.ETERNITY, VirtualColumns.EMPTY, Granularities.ALL, false, null);
+
+    final String exportFilePath = getExportFilePath(workerNumber, partitionNumber, exportFormat);
```

Comment:
This can be moved outside into the constructor, and needn't be called each time

Reply:
Moved